optaweb-employee-rostering-backend/src/main/java/org/optaweb/employeerostering/service/solver/EmployeeRosteringConstraintProvider.java

```diff
@@ -0,0 +1,347 @@
+                            return () -> container.decreaseHourlyCount(shift);
+                        },
+                        LoadBalancingHourCounter::getLoadBalance
+                ))
+                .penalizeConfigurableLong(CONSTRAINT_COVID_UNIFORM_DISTRIBUTION_OF_INOCULATED_HOURS,
+                        loadBalance -> loadBalance);
+    }
+
+    Constraint maximizeInoculatedHours(ConstraintFactory constraintFactory) {
+        return getAssignedShiftConstraintStream(constraintFactory)
+                .filter(shift -> shift.getEmployee().getCovidRiskType() == INOCULATED)
+                .rewardConfigurableLong(CONSTRAINT_COVID_MAXIMIZE_INOCULATED_HOURS, Shift::getLengthInMinutes);
+    }
+
+    Constraint migrationBetweenCovidAndNonCovidWards(ConstraintFactory constraintFactory) {
+        return getAssignedShiftConstraintStream(constraintFactory)
+                .filter(shift -> shift.getSpot().isCovidWard())
+                .join(Shift.class,
+                        equal(Shift::getEmployee),
+                        lessThanOrEqual(Shift::getEndDateTime, Shift::getStartDateTime))
+                .filter((covidShift, futureShift) -> !futureShift.getSpot().isCovidWard())
+                .ifNotExists(Shift.class,
+                        equal((covidShift, nonCovidShift) -> covidShift.getEmployee(), Shift::getEmployee),
+                        lessThanOrEqual((covidShift, nonCovidShift) -> covidShift.getEndDateTime(),
+                                Shift::getStartDateTime),
+                        greaterThanOrEqual((covidShift, nonCovidShift) -> nonCovidShift.getStartDateTime(),
+                                Shift::getEndDateTime))
+                .penalizeConfigurableLong(CONSTRAINT_COVID_MIGRATION_BETWEEN_COVID_AND_NON_COVID_WARDS,
+                        (covidShift, nonCovidShift) -> nonCovidShift.getLengthInMinutes());
+    }
+
+    Constraint requiredSkillForShift(ConstraintFactory constraintFactory) {
+        return getAssignedShiftConstraintStream(constraintFactory)
+                .filter(shift -> !shift.hasRequiredSkills())
+                .penalizeConfigurableLong(CONSTRAINT_REQUIRED_SKILL_FOR_A_SHIFT, Shift::getLengthInMinutes);
+    }
+
+    Constraint unavailableEmployeeTimeSlot(ConstraintFactory constraintFactory) {
+        return getConstraintStreamWithAvailabilityIntersections(constraintFactory, UNAVAILABLE)
+                .penalizeConfigurableLong(CONSTRAINT_UNAVAILABLE_TIME_SLOT_FOR_AN_EMPLOYEE,
+                        ((employeeAvailability, shift) -> shift.getLengthInMinutes()));
+    }
+
+    Constraint noOverlappingShifts(ConstraintFactory constraintFactory) {
+        return getAssignedShiftConstraintStream(constraintFactory)
+                .join(Shift.class,
+                        equal(Shift::getEmployee),
+                        lessThanOrEqual(Shift::getStartDateTime),
+                        lessThanOrEqual(Shift::getStartDateTime, Shift::getEndDateTime),
```

Comment:
Not sure if this is optimal; you are eliminating the `startTime = endTime` case in a filiter. Why not use `lessThan` and `greaterThan` instead of `lessThanOrEqual` and `greaterThanOrEqual`?

Reasoning:

Case 1: Shift A is completely before Shift B
then A.getEndDateTime() < B.getStartDateTime(), so no match

Case 2: Shift A ends when Shift B begins
then A.getEndDateTime() = B.getStartDateTime(), so no match (it a match currently)

Case 3: Shift A starts before Shift B, but ends within Shift B
then A.getEndDateTime > B.getStartDateTime and A.getStartDateTime < B.getEndDateTime, so match

Case 4: Shift A starts when Shift B starts, but ends before Shift B ends
then  A.getEndDateTime > B.getStartDateTime and A.getStartDateTime < B.getEndDateTime, so match

Case 5: Shift B is a subinterval of Shift A
then  A.getEndDateTime > B.getStartDateTime and A.getStartDateTime < B.getEndDateTime, so match

Case 6: Shift A starts when Shift B starts, and ends when Shift B ends
Then  A.getEndDateTime > B.getStartDateTime and A.getStartDateTime < B.getEndDateTime, so match

Case 7: Shift A starts when Shift B starts, and ends after Shift B ends
Then  A.getEndDateTime > B.getStartDateTime and A.getStartDateTime < B.getEndDateTime, so match

The first lessThanOrEqual eliminates all other cases